File: src/main/java/Day1/Test1.java
package Day1;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import scala.Tuple2;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;

public class Test1 {
    public static void main(String[] args) {
        SparkConf conf = new SparkConf().setAppName("RDD").setMaster("local[*]");
        JavaSparkContext sc = new JavaSparkContext(conf);
        JavaRDD<String> lines = sc.textFile(args[0]);
        //首先继续flatmap操作
        JavaRDD<String> flatMap = lines.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public Iterator<String> call(String s) throws Exception {
                return new ArrayList<String>(Arrays.asList(s.split(" "))).iterator();
            }
        });
        //这里为什么使用mapToPair  而不是 map
        JavaPairRDD<String, Integer> mapToPair = flatMap.mapToPair(new PairFunction<String, String, Integer>() {
            @Override
            public Tuple2<String, Integer> call(String s) throws Exception {
                return Tuple2.apply(s, 1);

            }
        });
        //对值继续sum
        JavaPairRDD<String, Integer> reduceByKey = mapToPair.reduceByKey(new Function2<Integer, Integer, Integer>() {
            @Override
            public Integer call(Integer v1, Integer v2) throws Exception {
                return v1 + v2;
            }
        });
        //排序,但是只能安key来,反转
        JavaPairRDD<Integer, String> mapToPair1 = reduceByKey.mapToPair(new PairFunction<Tuple2<String, Integer>, Integer, String>() {
            @Override
            public Tuple2<Integer, String> call(Tuple2<String, Integer> stringIntegerTuple2) throws Exception {
                return stringIntegerTuple2.swap();


            }
        });
        //排序
        JavaPairRDD<Integer, String> sortByKey = mapToPair1.sortByKey(false);
        //再次反转
        JavaPairRDD<String, Integer> mapToPair2 = sortByKey.mapToPair(new PairFunction<Tuple2<Integer, String>, String, Integer>() {
            @Override
            public Tuple2<String, Integer> call(Tuple2<Integer, String> integerStringTuple2) throws Exception {
                return integerStringTuple2.swap();
            }
        });

        mapToPair2.saveAsTextFile(args[1]);
        sc.stop();
    }
}

File: src/main/scala/cn/tongyongtao/Day15/Test1.scala
package cn.tongyongtao.Day15

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

//数据如下类型,需求获取每个月份的总金额并且迭代增加
//1,2020-01-02,20
//1,2020-01-03,20
//1,2020-01-04,20
object Test1 {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setAppName(this.getClass.getSimpleName)
    if (args(0).toBoolean) {
      conf.setMaster("local[*]")
    }
    val sc = new SparkContext(conf)
    //数据的结构: ((1,2020-1),20)
    val rdd = sc.textFile(args(1))
    val maprdd = rdd.map(data => {
      val txt = data.split(",")

      val shopid = txt(0)
      val date = txt(1).split("-")
      val turnover = txt(2).toInt
      ((shopid, (date(0) + "-" + date(1))), turnover)
    })
    val reducebyrdd: RDD[((String, String), Int)] = maprdd.reduceByKey(_ + _)
    //是否可以避免掉??
    val re = reducebyrdd.map({
      case ((x, y), z) => ((x), (y, z))
    }).groupByKey().map(data => {
      //使用一个变量来接收一个累加值
      val sorted: List[(String, Int)] = data._2.toList.sorted
      var sum = 0
      sorted.map(num => {
        sum += num._2
        (data._1, (num._1, sum))
      })
    })
    re.collect().foreach(println)
    sc.stop()
  }
}
